pkg/controllers/provisioningv2/rke2/provisioningcluster/template.go

```diff
@@ -287,6 +287,8 @@ func machineDeployments(cluster *rancherv1.Cluster, capiCluster *capi.Cluster, d
 				Name:        machinePoolName,
 				Labels:      machineDeploymentLabels,
 				Annotations: machinePool.MachineDeploymentAnnotations,
+				// This ensures that the CAPI cluster is not deleted before the machines are deleted.
+				Finalizers: []string{metav1.FinalizerDeleteDependents},
```

Comment:
How does this block CAPI cluster deletion if we are adding the finalizer to the `MachineDeployment`

Reply:
I added it here because if the MachineDeployment has it then:
1. The MachineDeployment CAPI controller will put in on the MachineSet object. Then, since the MachineDeployment object has it, it will not be removed until all the MachineSet objects it owns are deleted.
2. If the MachineSet object has it then the MachineSet object won’t be removed until all the Machine objects are deleted.
3. The Machine objects will wait for all the infrastructure machine objects to be cleaned up before the Machine object gets removed.

This blocks the CAPI cluster object from being deleted because the controller will loop, look for MachineDeployments that it owns. If it finds any, it will wait to remove its finalizer. If it doesn’t find any, then the controller will remove the finalizer and the cluster object will delete.